### PR TITLE
# Handle masked pixels (UNSEEN / NaN) in spherical harmonic transforms

### PR DESCRIPTION
**Pull request by ArtemBasyrov**
## Summary

This PR makes `map2alm`, `map2alm_spin`, and `smoothing` handle masked and invalid pixels the way healpy does, so that maps containing `UNSEEN` or non-finite values produce well-defined results instead of silently corrupted ones.

## Motivation

Previously the transforms passed pixel data straight to the s2fft backend with no masking. A single `UNSEEN` pixel (the HEALPix bad-value sentinel, `-1.6375e30`) is a huge finite number, so it dominated every output coefficient; a single `NaN`/`inf` propagated into the whole result. The `map2alm` docstring already promised healpy-style `UNSEEN` handling, but the code did not implement it. This PR closes that gap and extends it to non-finite values.

## What changed

`map2alm` and `map2alm_spin` now zero out bad pixels before the transform, on a copy that leaves the input map untouched. Each map in a stack is masked independently. "Bad" means `UNSEEN` (detected with a tolerant comparison) or non-finite (`NaN` / `+-inf`).

`smoothing` now remembers the bad pixels of each input map and restores them to `UNSEEN` at the same positions in the smoothed output, matching healpy's round-trip behavior. Clean maps are unaffected.

`mask_bad` (in `pixelfunc`) was upgraded from an exact `m == UNSEEN` comparison to healpy's tolerant comparison `|m - badval| <= atol + rtol * |badval|`, with the same `badval` / `rtol` / `atol` signature and defaults. It is now exported from the top-level package.

## Behavior vs healpy

For `UNSEEN`, the output matches healpy to machine precision (round-trip parity at ~1e-14, and the same set of restored pixels in `smoothing`). The one deliberate extension beyond healpy is non-finite handling: healpy's `mask_bad` does not flag `NaN`/`inf`, so in healpy a single `NaN` poisons the entire output; here `NaN`/`inf` pixels are treated as bad and zeroed (and, in `smoothing`, restored as `UNSEEN`).

`alm2map` is intentionally unchanged, because its input is harmonic coefficients and its output is a full-sky band-limited map, so there are no masked pixels to handle there; the round trip is fixed entirely on the analysis side.

## Testing

Added tests covering: `UNSEEN` parity against healpy, equivalence to manually zeroing the masked pixels, input map not being mutated, `NaN`/`inf` handling, independent masks across I/Q/U, tolerant `mask_bad`, and `smoothing` restoring the mask (including the single-map polarized case). All new tests pass; `_sanitize_map` is verified to be a bit-exact no-op on maps with no bad pixels.

### CODE REVIEW

File: jax_healpy/sphtfunc.py [hunk 1]
```diff
@@ -1630,14 +1648,17 @@ def smoothing(
         alms_smooth = smoothalm(
             alms, fwhm=fwhm, sigma=sigma, beam_window=beam_window, pol=True, mmax=mmax, healpy_ordering=False
         )
-        return alm2map(alms_smooth, nside=nside, lmax=lmax, mmax=mmax, pol=True, healpy_ordering=False)
+        out = alm2map(alms_smooth, nside=nside, lmax=lmax, mmax=mmax, pol=True, healpy_ordering=False)
+        # A single (1, npix) input collapses to a (npix,) map; align the mask shape.
+        return jnp.where(bad.reshape(out.shape), UNSEEN, out)
```
Comment (ASKabalan) [inline jax_healpy/sphtfunc.py:1653]: I don't understand the need for reshaping
if bad is npix size and out is B , npix
The batching should take care of this

File: jax_healpy/sphtfunc.py [hunk 2]
```diff
@@ -1630,14 +1648,17 @@ def smoothing(
...
 
     # Scalar branch (single map).
     alms = map2alm(map_in, lmax=lmax, mmax=mmax, iter=iter, pol=False, healpy_ordering=False)
     alms_smooth = smoothalm(
         alms, fwhm=fwhm, sigma=sigma, beam_window=beam_window, pol=False, mmax=mmax, healpy_ordering=False
     )
-    return alm2map(alms_smooth, nside=npix2nside(map_in.shape[-1]), lmax=lmax, mmax=mmax, healpy_ordering=False)
+    out = alm2map(alms_smooth, nside=npix2nside(map_in.shape[-1]), lmax=lmax, mmax=mmax, healpy_ordering=False)
+    return jnp.where(bad.reshape(out.shape), UNSEEN, out)
```
Comment (ASKabalan) [inline jax_healpy/sphtfunc.py:1661]: same as above

**Review by ASKabalan — COMMENTED** [review]
Seems god
Only the reshaping thing is not justified maybe
reshaping can some times destroy sharding information 
in this case it is highly unlikely
but since the bad mask is make by you and it is guarenteed to be shape npix
Then why does it have to be reshaped to match the output?
